src/tools/piv-tool.c

```diff
@@ -603,19 +596,25 @@ int main(int argc, char *argv[])
 	if (action_count <= 0)
 		goto end;
 
-	if (opt_driver != NULL) {
-		err = sc_set_card_driver(ctx, opt_driver);
-		if (err) {
-			fprintf(stderr, "Driver '%s' not found!\n", opt_driver);
-			err = 1;
-			goto end;
-		}
+	/* force PIV card driver */
+	err = sc_set_card_driver(ctx, "PIV-II");
+	if (err) {
+		fprintf(stderr, "PIV card driver not found!\n");
+		err = 1;
+		goto end;
 	}
 
 	err = util_connect_card(ctx, &card, opt_reader, opt_wait, verbose);
 	if (err)
 		goto end;
 
+	/* fail if card is not a PIV card */
+	if (card->type < SC_CARD_TYPE_PIV_II_BASE || card->type >= SC_CARD_TYPE_PIV_II_BASE+1000) {
+		fprintf(stderr, "Card type %X: not a PIV card\n", card->type);
+		err = 1;
+		goto end;
+	}
+
```

Comment:
I think you don't need to check the type of the driver if you force it with `sc_set_card_driver()`. Rather, you should print the error that "no compatible PIV card has been found"  above where `util_connect_card()` cannot connect to a card.

(Same for dnie-tool and cardos-tool...)

Reply:
Sorry I need to disagree.

At least the `dnie` driver will happily connect to the card [i.e. `util_connect_card()` succeeds], and spew unknown APDUs at  unsuspecting cards of any type if this check is left out. 
As the additional check does not hurt if `util_connect_card()` fails, I rather opt for safety here.

Please commit.

Reply:
util_connect_card calls sc_connect_card, which calls dnie_init, which errors if a card doesn't support the DNIe commands, doesn't it?

Reply:
I do not know about the internals of the `dnie` driver.
But when I compiled `dnie-tool`and ran it against my Yubikey (which definitely is no DNIe card), `util_connect_card()` definitely succeeded and without the additional check, `dnie-tool` spat it's APDUs to my Yubikey.
That's why I think the checks are necessary. In any case they do not hurt.

Please merge!

Reply:
Another approach is to set the default value  for `opt_driver` to the name of the driver for card specific tools.  So piv-tool.c would have `const char *opt_driver = "PIV-II";` and dnie-tool.c would have ` const char *opt_driver = "dnie";` User could still override this for what ever reason. But only to appropriate driver would be used. 

Six tools support `--card-driver` which have `static const char *opt_driver = NULL;`  So another approach in the card specific drivers: `dnie-tool.c`, `cardos-tool.c` `piv-tool-c` and `westcos-tool.c`  change to something like: `static const char *opt_driver = "dnie";` or `const char *opt_driver = "PIV-II";`. User could still use the `--card-driver` if they wanted too. 

The same could be done in tools that do not have `--card-driver` option but are card specific. These are: `cryptoflex-tool.c`, `egk-tool.c`, `egk-tool-cmdline.c`, `gids-tool.c`, `goid-tool.c`, `goid-tool-cmdline.c`, `iasecc-tool.c`, `netkey-tool.c`, `npa-tool.c`, `npa-tool-cmdline.c` and `openpgp-tool.c`. 

If you go with testing the card type, the above list of card specific tools should check for their card types too.

Reply:
@dengert of course there are many ways how to do it. 
Frank asked me to extend the original patch, which had very little to do with the topics at hand, in one specific way. Which I did.
I am willing to write another patch, but only after this one gets merged and you and Frank agree on the approach.

Reply:
npa-tool and goid-tool don't rely on a specific card driver.

@dengert feel free to set the card driver as well in the other tools you mentioned. The code in this PR is a good starting point.